saleor/graphql/warehouse/types.py

```diff
@@ -0,0 +1,80 @@
+import graphene
+
+from ...core.permissions import StockPermissions
+from ...warehouse import models
+from ...warehouse.availability import get_available_quantity_for_customer
+from ..account.enums import CountryCodeEnum
+from ..core.connection import CountableDjangoObjectType
+from ..decorators import permission_required
+
+
+class WarehouseAddressInput(graphene.InputObjectType):
+    street_address_1 = graphene.String(description="Address.", required=True)
+    street_address_2 = graphene.String(description="Address.")
+    city = graphene.String(description="City.", required=True)
+    city_area = graphene.String(description="District.")
+    postal_code = graphene.String(description="Postal code.")
+    country = CountryCodeEnum(description="Country.", required=True)
+    country_area = graphene.String(description="State or province.")
+    phone = graphene.String(description="Phone number.")
+
+
+class WarehouseInput(graphene.InputObjectType):
+    name = graphene.String(description="Warehouse name.", required=True)
+    company_name = graphene.String(description="Company name.")
+    shipping_zones = graphene.List(
+        graphene.ID, description="Shipping zones supported by the warehouse."
+    )
+    email = graphene.String(description="The email address of the warehouse.")
+
+
+class WarehouseCreateInput(WarehouseInput):
+    address = WarehouseAddressInput(
+        description="Address of the warehouse.", required=True
+    )
+
+
+class WarehouseUpdateInput(WarehouseInput):
+    address = WarehouseAddressInput(
+        description="Address of the warehouse.", required=False
+    )
+
+
+class Warehouse(CountableDjangoObjectType):
+    class Meta:
+        description = "Represents warehouse."
+        model = models.Warehouse
+        interfaces = [graphene.relay.Node]
+        only_fields = [
+            "id",
+            "name",
+            "company_name",
+            "shipping_zones",
+            "address",
+            "email",
+        ]
+
+
+class StockInput(graphene.InputObjectType):
+    product_variant = graphene.ID(
+        required=True, description="Product variant assiociated with stock."
+    )
+    warehouse = graphene.ID(
+        required=True, description="Warehouse in which stock is lockated."
+    )
+    quantity = graphene.Int(description="Quantity of items available for sell.")
+
+
+class Stock(CountableDjangoObjectType):
+    available = graphene.Int(description="Quantity of a product available for sale.")
```

Comment:
This type exposes three quantities:

```
  quantity: Int!
  quantityAllocated: Int!
  available: Int
```

It's really important that we provide meaningful names and descriptions here, otherwise, it will be really confusing to understand and use.

- We need to override `quantity` and `quantity_allocated` fields in this Graphene type so that we can provide descriptions. I would take descriptions from the fields that this PR deprecates in the `ProductVariant` type.
- the field `available` is nullable in it the schema - is it on purpose? I think we will always have an integer value here.
- if I'm correct the field `available` does what `ProductVariant.stockQuantity` used to do. Can we use the same name for this field then?